move network + dns container args to pureconfig

tysonnorris opened the pull request:
also add "extra" container args to pureconfig

The main motivation here was to allow all action containers to some additional env vars (same value used for all containers). To get there, I moved network + dns args to pureconfig as well. 

There are still more container args that could be refactored, or it would be possible to move everything to the "extra-args" config.  These are left as an exercise for another time.

markusthoemmes commented on common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala:
```diff
@@ -57,3 +65,23 @@ trait ContainerFactoryProvider extends Spi {
+  val containerArgsConfig = {
+    //since typesafe (lightbend) Config does NOT support parsing complex objects from system properties, we need a ConfigReader
+    //see https://github.com/lightbend/config/blob/master/HOCON.md#java-properties-mapping for details on properties handling from typesafe Config
+    //see pureconfig docs at https://pureconfig.github.io/docs/combinators.html
```
Can you elaborate on this a bit?

tried this locally:

```scala
case class ContainerArgsConfig(network: String, dnsServers: Seq[String], extraArgs: Map[String, Set[String]])

object Testing {
  import pureconfig._

  def main(args: Array[String]): Unit = {
    val config = """
                   |network: bridge
                   |dns-servers: [google.com]
                   |extra-args: {
                   |  test: [a, b]
                   |  another: [c, d]
                   |}
                   | """.stripMargin

    println(loadConfigOrThrow[ContainerArgsConfig](ConfigFactory.parseString(config)))
  }
}
```

Which yields:

```
ContainerArgsConfig(bridge,List(google.com),Map(another -> Set(c, d), test -> Set(a, b)))
```

Do we need something different?

tysonnorris replied:
I tried to be as specific as possible in the comment :)

It is this: _typesafe (lightbend) Config does NOT support parsing complex objects from system properties_ i.e. when you override values via system properties (e.g. the `CONFIG_*` env vars in our case), they are always read as String type, never converted to List/Map/Object/etc.

So, in your example, you did not override the config from system properties.

IMHO this is a real flaw in lightbend/typesafe config, and we just need to be very careful about what types are parsed from the config, and make sure the override via system property works as you expect. I saw some similar logged issues with lightbend config, along with some arguments as to why it won't be fixed...

In fact, I just notice the same flaw exists for the `dns-servers` property - I'll add a config reader for that as well. Now I'm also wonder if the DockerClientTimeoutConfig will exhibit the same issue, if you try to override via system properties.

tysonnorris replied:
Actually for 'dns-servers' (type `Seq[String]`), there exists a way to handle this via typesafe java properties parser: https://github.com/lightbend/config/pull/70

You have to specify each element in the list as a separate system property whose key ends with the index number in the list, e.g.
```
    System.setProperty("whisk.container-factory.container-args.dns-servers.0", "google.com")
    System.setProperty("whisk.container-factory.container-args.dns-servers.1", "1.2.3.4")
```

Note that we _could_ alternatively choose to add a custom parser for this, so that we can use a single property like "dns-servers"="[google.com,1.2.3.4]", but I ran into this issue trying to use the typesafe parser https://github.com/lightbend/config/issues/460 so we'd need to parse the value in yet another way.

markusthoemmes replied:
Got it, that makes sense!

Actually using your workaround for the lists, overriding the extra args works exactly the same way:

```scala
case class ContainerArgsConfig(network: String, dnsServers: Seq[String], extraArgs: Map[String, Set[String]])

object Testing {
  import pureconfig._

  def main(args: Array[String]): Unit = {
    System.setProperty("network", "test")
    System.setProperty("dns-servers.0", "google.com")
    System.setProperty("extra-args.test.0", "a")
    System.setProperty("extra-args.test.1", "b")
    System.setProperty("extra-args.try.0", "a")

    println(loadConfigOrThrow[ContainerArgsConfig](""))
  }
}
```

markusthoemmes commented on common/scala/src/main/scala/whisk/core/containerpool/ContainerFactory.scala:
```diff
@@ -18,15 +18,23 @@
+import com.typesafe.config.ConfigFactory
+import java.util.ArrayList
+import pureconfig.ConfigReader
+import pureconfig.loadConfigOrThrow
```
You want to import `pureconfig._` here to get all the implicits.

mcdan commented on core/invoker/src/main/resources/application.conf:
```diff
@@ -19,4 +19,12 @@ whisk {
   }
+
+  # args for 'docker run' to use
+  container-factory.container-args {
```
This is great I can add the container root path from #3196 to this struct instead of where I put it.

tysonnorris replied:
There is a subtle difference here in that the container-args config is not something that is specific to _docker cli_, rather it will apply to non-docker ContainerFactory impls (mesos/kube), while the container root path is very specific to docker cli. 
So I would keep the configs separated to:
- `whisk.docker`
- `whisk.container-factory`

markusthoemmes commented on core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala:
```diff
@@ -83,7 +85,7 @@ class InvokerReactive(config: WhiskConfig, instance: InstanceId, producer: Messa
           "--ulimit" -> Set("nofile=1024:1024"),
           "--pids-limit" -> Set("1024"),
-          "--dns" -> config.invokerContainerDns.toSet) ++ logsProvider.containerParameters)
+          "--dns" -> containerArgs.dnsServers.toSet) ++ logsProvider.containerParameters)
```
I'm wondering: The dns configuration seems duplicated (given it's set in the factory itself again as well). Can we remove it here and thus remove the dependency on the config from `InvokerReactive`?

tysonnorris replied:
well, now I'm noticing a problem that _transformEnvironment.sh_ is ignoring my env vars that end with index, e.g. `CONFIG_whisk_containerFactory_containerArgs_dnsServers.0=8.8.8.8`; will look into that...

tysonnorris replied:
oh, using `_<index>` works...

tysonnorris commented:
@markusthoemmes I updated to remove ContainerArgsConfig from InvokerReactive, and also verified that the dnsServers config is adjustable from env vars. 

In order to leave the existing ansible config for dns functional, I had to add some changes to the deploy.yml for invoker, specifically:
```        {% for item in (invoker_container_network_dns_servers | default()).split(' ')  %}
        -e CONFIG_whisk_containerFactory_containerArgs_dnsServers_{{loop.index0}}={{ item }}
        {% endfor %}
```
Let me know if you have concerns. Tested locally fine, FWIW.

tysonnorris commented:
Any concerns on this @markusthoemmes @rabbah ?

After this is merged I'll take a stab at fixing #3110 along with moving some (all?) of the other invoker configs to pureconfig.

markusthoemmes commented:
PR1 2545 🔵

chetanmeh commented on tests/src/test/scala/whisk/core/containerpool/test/ContainerArgsConfigTest.scala:
```diff
@@ -0,0 +1,58 @@
+  }
+
+  it should "override defaults from system properties" in {
+    System.setProperty("whisk.container-factory.container-args.extra-args.label.0", "l1")
```
May be the tests should revert changes in System property post run otherwise it may lead to side effects in other test

tysonnorris replied:
agreed - since this just got merged, I will address in a future pr - this area will get some revisions as I work on some of the containerfactory details for mesos
